### PR TITLE
Template `distribute_entity_data` to support multiple data-types

### PR DESCRIPTION
**Pull request by jorgensd**
First step towards resolving https://github.com/FEniCS/dolfinx/issues/1253
Second step would be to template `XDMFFile.read_meshtags`. This is fairly trivial, but I figured we should agree on the core changes before doing that change.

Currently tested with:
```python
from mpi4py import MPI
import numpy as np
import dolfinx

mesh = dolfinx.mesh.create_unit_square(MPI.COMM_WORLD, 2, 2)
num_vertices = mesh.topology.index_map(0).size_local
entities = np.arange(num_vertices, dtype=np.int32)


def distribute_data(dtype):
    values = entities.astype(dtype).copy()
    if dtype == np.float64:
        values -= 0.6
    return dolfinx.io.distribute_entity_data(mesh, 0, entities.reshape(-1, 1).copy(), values[::-1].copy())


a = distribute_data(np.int32)

b = distribute_data(np.float64)
c = distribute_data(np.complex128)
print(a, b, c)
```
I've removed the unnecessary mesh geometry templating of this function, and made the Python-binding more low-level (improving the Python layer of `distribute_entity_data` to work with the new binding). The PR also has documentation improvements of this function.

This would enabling reading vertex, facet and cell data that is single or double precision, which is the case in many biomedical applications (where one for instance have voxel data, or material parameters).

The major difference in the code is that the communication of entity indices and values are split into two separate neighbourhood alltoallv calls. 

No major changes in the actual functional structure of the code.

### CODE REVIEW

**Comment by jorgensd** [comment]
@massimiliano-leoni could you test this?:)

File: python/dolfinx/wrappers/io.cpp [hunk 1]
```diff
@@ -349,7 +360,10 @@ void io(nb::module_& m)
   declare_vtx_writer<float>(m, "float32");
   declare_vtx_writer<double>(m, "float64");
 
-  declare_real_types<float>(m);
-  declare_real_types<double>(m);
+  declare_data_types<std::int32_t>(m);
+  // declare_data_types<float>(m);
```
Comment (jorgensd) [inline python/dolfinx/wrappers/io.cpp:364]:
```suggestion
  declare_data_types<float>(m);
```

File: cpp/dolfinx/io/xdmf_utils.h [hunk 2]
```diff
@@ -24,6 +26,13 @@
 #include <utility>
 #include <vector>
 
+namespace
```
Comment (garth-wells) [inline cpp/dolfinx/io/xdmf_utils.h:29]: Anonymous namespaces should never appear in header files - should only appear in `.cpp` files.

**Review by garth-wells — CHANGES_REQUESTED** [review]
I would suggest leaving the code in the `.cpp` file and explicitly instantiating for the small number of types that we'll use (at this stage). It's preferable to keep code in the implementation file. This is what we do elsewhere for the mesh.

File: cpp/dolfinx/io/xdmf_utils.h [hunk 3]
```diff
@@ -15,6 +15,8 @@
 #include <concepts>
 #include <dolfinx/common/MPI.h>
 #include <dolfinx/common/types.h>
+#include <dolfinx/fem/ElementDofLayout.h>
+#include <dolfinx/mesh/Topology.h>
```
Comment (jorgensd) [inline cpp/dolfinx/io/xdmf_utils.h:19]:
```suggestion
```

**Comment by jorgensd** [comment]
@garth-wells any further changes you would like to see in this PR?

**Review by garth-wells — APPROVED** [review]

**Pull request merged**